nautobot/dcim/forms.py

```diff
@@ -1982,6 +2173,191 @@ class DeviceFilterForm(
+        queryset=Manufacturer.objects.all(),
+        required=False,
+        initial_params={"module_types": "$module_type"},
+    )
+    module_type = DynamicModelChoiceField(
+        queryset=ModuleType.objects.all(),
+        query_params={"manufacturer": "$manufacturer"},
+    )
+    parent_module_bay_device_filter = DynamicModelChoiceField(
+        queryset=Device.objects.all(),
+        required=False,
+        label="Parent Device",
+        query_params={"has_empty_module_bays": True},
+        initial_params={"module_bays": "$parent_module_bay"},
+    )
+    parent_module_bay_device = DynamicModelChoiceField(
+        queryset=ModuleBay.objects.all(),
+        label="Parent Module Bay",
+        required=False,
+        query_params={"parent_device": "$parent_module_bay_device_filter", "has_installed_module": False},
+        initial_params={"pk": "$parent_module_bay", "parent_device__module_bays": "$parent_module_bay"},
+    )
+    parent_module_bay_module_filter = DynamicModelChoiceField(
+        queryset=Module.objects.all(),
+        required=False,
+        label="Parent Module",
+        query_params={"has_empty_module_bays": True},
+        initial_params={"module_bays": "$parent_module_bay"},
+    )
+    parent_module_bay_module = DynamicModelChoiceField(
+        queryset=ModuleBay.objects.all(),
+        label="Parent Module Bay",
+        required=False,
+        query_params={"parent_module": "$parent_module_bay_module_filter", "has_installed_module": False},
+        initial_params={"pk": "$parent_module_bay", "parent_module__module_bays": "$parent_module_bay"},
+    )
+    location = DynamicModelChoiceField(
+        queryset=Location.objects.all(),
+        required=False,
+        label="Location",
+        query_params={"content_type": Module._meta.label_lower},
+    )
+    role = DynamicModelChoiceField(
+        queryset=Role.objects.all(),
+        required=False,
+        query_params={"content_types": Module._meta.label_lower},
+    )
+
+    class Meta:
+        model = Module
+        fields = [
+            "manufacturer",
+            "module_type",
+            "parent_module_bay",
+            "location",
+            "serial",
+            "asset_tag",
+            "role",
+            "status",
+            "tenant_group",
+            "tenant",
+            "tags",
+        ]
+        help_texts = {
+            "serial": "Module serial number",
+        }
+
+    def clean(self):
+        cleaned_data = self.cleaned_data
+        if cleaned_data["parent_module_bay_device"] and cleaned_data["parent_module_bay_module"]:
+            raise forms.ValidationError("Multiple parent module bays selected.")
+        elif cleaned_data["parent_module_bay_device"]:
+            cleaned_data["parent_module_bay"] = cleaned_data.pop("parent_module_bay_device")
+        elif cleaned_data["parent_module_bay_module"]:
+            cleaned_data["parent_module_bay"] = cleaned_data.pop("parent_module_bay_module")
```

Comment:
I wonder if it'd make sense, alternately, for `clean()` to throw away any `location` value if a `parent_module_bay` is also specified, just to avoid a likely user error/misunderstanding? Or do you think it's better for this to be a ValidationError that the user needs to address?